sklearn/datasets/_samples_generator.py

```diff
@@ -1239,12 +1240,20 @@ def make_low_rank_matrix(
     return np.dot(np.dot(u, s), v.T)
 
 
+# TODO: Change argument `data_transposed` default from True to False in 1.3.
+# TODO: Deprecate data_transposed in 1.3, always return data not transposed.
 def make_sparse_coded_signal(
-    n_samples, *, n_components, n_features, n_nonzero_coefs, random_state=None
+    n_samples,
+    *,
+    n_components,
+    n_features,
+    n_nonzero_coefs,
+    random_state=None,
+    data_transposed="warn",
 ):
     """Generate a signal as a sparse combination of dictionary elements.
 
-    Returns a matrix Y = DX, such as D is (n_features, n_components),
+    Returns a matrix Y = DX, such that D is (n_features, n_components),
```

Comment:
With `data_transposed=True`, we get this relationship:

```python
Y, D, X = make_sparse_coded_signal(..., data_transposed=True)
# Y = D @ X
```

With `data_transposed=False`, the relationship is flipped:

```python
Y, D, X = make_sparse_coded_signal(..., data_transposed=False)
# Y = X @ D
```

If we want this behavior, then I think we need to document this equality when `data_transposed=False`.

Another option is to actually change the order when `data_transposed=False`:

```python
Y, X, D = make_sparse_coded_signal(..., data_transposed=False)
# Y = X @ D
```

(I do not like doing this, but the parameter does say "transpose" and we are transposing a matrix multiply, so it kind of makes sense)

@jeremiedbb Was this discussed somewhere before?

Reply:
> @jeremiedbb Was this discussed somewhere before?

I don't think so. I guess the idea to just return transposed matrices is that the order of the returned objects does not change based on an argument. I'm +0 on changing the order. Once we have deprecated the initial behavior there should not be confusion any more. But I agree that the correct operation should be documented.

(the names are pretty bad imo and that's the main source of confusion to me, Y is what we call X basically everywhere and X is the code)

Reply:
@thomasjpfan Thank you so much for reviewing the changes. Please let me know if there is anything I can do to help.

Reply:
Looking at this again, if `data_transposed=False` becomes the default, I prefer changing the order, i.e:

```python
Y, X, D = make_sparse_coded_signal(..., data_transposed=False)
# Y = X @ D
```

I think we need another opinion to move forward.

Reply:
no, no, it's fine, let's go with changing the order. The sooner we deprecate, the better :)